File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        n = 0
        node = head
        while node:
            n += 1
            node = node.next
        
        mid = n // 2 + 1
        n = 1
        node = head
        while n < mid:
            n += 1
            node = node.next
        
        return node
        

File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        if (head.next == null) {
            return head;
        }
        
        // calculate list's length
        int len = 0;
        ListNode node = head;
        while(node != null) {
            len++;
            node=node.next;
        }
        
        // System.out.println(len);
        int target = len / 2 + 1;
        int cnt = 1;
        node = head;
        while (cnt < target) {
            cnt++;
            node = node.next;
        }
        
        
        return node;
    }
}